source/common/thread_local/thread_local_impl.cc

```diff
@@ -60,6 +99,32 @@ void InstanceImpl::registerThread(Event::Dispatcher& dispatcher, bool main_threa
     dispatcher.post([&dispatcher] { thread_local_data_.dispatcher_ = &dispatcher; });
   }
 }
+void InstanceImpl::recycle(std::unique_ptr<SlotHolder>&& holder) {
+  if (holder->isRecycleable()) {
+    holder.reset();
+    return;
+  }
+  deferred_deletes_.emplace_back(std::move(holder));
```

Comment:
Nothing tracks when this holder is eventually recyclable, right? Meaning, in this case it will stick around the deferred delete list until something else gets scheduled for cleanup? (Since presumably this schedule cleanup can race with the thing actually being recyclable). Maybe this is fine? Or maybe you should be checking for use count on the worker threads and scheduling cleanup if its detached and ready to be recycled (the last one)?

Reply:
I think it's tracked: 
1. A cleanup task is scheduled when we put a new item into the deferred delete queue.
2. The cleanup task will cleanup recyclable items from the Q, and reschedule another cleanup task(the recursion) if there are any non-recyclable ones. 

So once an item is put into the queue, the above event-loop (1->2->2->2->....) will guarantee the items been cleaned up.

Reply:
> Stepping back though, can you refresh my memory on why this complexity is needed vs. just cancelling any potential updates before we start listener shutdown which I think would be serialized correctly? Thank you!
> 
> /wait

I think the canceling/waiting approach requires more changes to the code base. Basically, if we want to be lock free, we'd have to split the LDS update (or Server shutdown) into 2 phases: 
1. set an flag or broadcast canceling signals to workers.
2. Wait to be signaled that the 'shutdown' message has been received by everyone.

In the meantime, if it's a LDS update caused ListenerImpl teardown, we have to bookkeep all the TLS resources owned by "this" ListenerImpl, and only wait on them to be cleaned up, as a RDS subscription (or any other subscription is shared between Listeners). 

On the other hand, this change limit changes to tls module only, least disturbance to the code base.

Reply:
> Basically, if we want to be lock free, we'd have to split the LDS update (or Server shutdown) into 2 phases

To be clear, this is how it already works. See `ListenerManagerImpl::drainListener` which already handles draining and stopping a listener, including posting a completion back and forth. Assuming all xDS updates are stopped before removal (or drain) by definition there can be no destruction race condition. What I'm trying to get at is how hard is it really to just shut down LDS/RDS/etc. updates for a listener on the main thread before doing this? Naively it doesn't seem that hard.

Reply:
I talked to @stevenzzzz offline and we are going to time box trying to implement explicit shutdown of resources prior to listener shutdown. This might be accomplished via the init manager or some similar structure. If it ends up being too hard or has complications we can come back to this.

/wait

Reply:
I'd be in favor of not repeating something like init manager. Init manager is has turned out tobe hard to reason about, with a maze of callbacks, watchers, registration conditions, lifetime considerations and so on. History speaks for itself with all the bugs and races on this one.

I think there is a good argument to be made that the solution in this PR has benefits in that it simplifies the contract around slot ownership, e.g. the `runOnAllThreads()` gets the threads out of the business of worrying about `tls_` slow ownership and lifetime. It also centralizes all the complexity inside of the `SlotImpl`, whereas init manager style solutions diffuse this complexity.

So, my $0.02 is that something like the solution in this PR is a nice direction; basically getting away from having threads worrying about TLS slots directly.

Reply:
Throwing out another idea that has come up when chatting with @stevenzzzz; we could have a centralized (owned by `Server::Instance`) TLS slot manager. This will be the actual owner of the memory backing all slots and it lives forever. It would hand out slot handles and then have them returned via RAII.

Reply:
> So, my $0.02 is that something like the solution in this PR is a nice direction; basically getting away from having threads worrying about TLS slots directly.

I think with this solution there are going to be potentially hard to reason about thread interleavings, and I have no doubt we will have a different set of issues from this direction, though I agree it's a nice solution and it will work in the end. My reason for proposing the use of the existing init manager system is that it already exists. We must register things into the init manager, and I think it would be relatively trivial to also register a shutdown hook of some type at the same time.

With that said, I don't feel that strongly about it, so if you both want to continue with this solution we can do that.

> we could have a centralized (owned by Server::Instance) TLS slot manager.

This is the way it already works?

Reply:
Good discussion here.
Personally I am not a fan of the init-manager mechanism. It's essentially a barrier that's passed around and hard to track what's been barriered and what's(should) not. 

The other problem I want to call out here is that no matter which way we go, we need to let contributors know about this problem, and we should not distributed (by capturing some shared_ptr) ownership of a tls slot or its owner to worker threads. 
That has the side effect of destructing the delicate dependency graph of objects in ServerImpl and cause crash during server shutdown (IIUC, per my experience though we drain the libevent message queue when shutdown a worker, we don't prevent new callbacks been posted to Dispatcher's callback queue). An example would be post a 
a callback which captures a shared_ptr<RdsRouteConfigProvider> to workers. On destruction, The singletonManager is destructed first, and then the RdsConfigSubscription owned by RdsRouteConfigProvider will try to remove it self from the destructed RdsConfigProviderManager.

Reply:
> The other problem I want to call out here is that no matter which way we go, we need to let contributors know about this problem, and we should not distributed (by capturing some shared_ptr) ownership of a tls slot or its owner to worker threads.

This is my point exactly. Although I agree Init/ShutdownManager is problematic in certain ways, its extremely explicit on what needs to happen. Basically, don't do any more updates. We are trading that for rules around how this can be used which are hard to understand and enforce. With that, I will stop there. @htuch I assume you want to move forward with this solution?